miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultContainerService.java

```diff
@@ -122,6 +129,32 @@ private void validateChange(SequencerPartitionContainer container, SequencerPart
       errors.add(new ValidationError("identificationBarcode", "There is already a container with this serial number"));
     }
 
+    if (beforeChange != null && ValidationUtils.isSetAndChanged(SequencerPartitionContainer::getModel, container, beforeChange)) {
+      SequencingContainerModel before = beforeChange.getModel();
+      SequencingContainerModel after = container.getModel();
+      if (after.getPlatformType() != before.getPlatformType()) {
+        errors.add(new ValidationError("model.id",
+            String.format("Can only be changed to a model of the same platform (%s)", before.getPlatformType().getKey())));
+      } else if (after.getPartitionCount() != before.getPartitionCount()) {
+        errors.add(new ValidationError("model.id",
+            String.format("Can only be changed to a model with the same number of partitions (%d)", before.getPartitionCount())));
+      }
+      if (beforeChange.getRunPositions() != null) {
+        Stream<InstrumentModel> requiredInstrumentModels = beforeChange.getRunPositions().stream()
+            .map(RunPosition::getRun)
+            .map(Run::getSequencer)
+            .map(Instrument::getInstrumentModel);
+        if (requiredInstrumentModels.anyMatch(required -> after.getInstrumentModels().stream()
+            .map(InstrumentModel::getId)
+            .noneMatch(id -> id == required.getId()))) {
+          errors.add(new ValidationError("model.id",
+              String.format("Can only change to a model compatible with the linked runs' instrument models (%s)",
+                  LimsUtils.joinWithConjunction(requiredInstrumentModels.map(InstrumentModel::getAlias).collect(Collectors.toSet()),
```

Comment:
Streams aren't supposed to be reused. Maybe create a list or set of instrument models?

Reply:
fixed